lib/iris/_shapefiles.py

```diff
@@ -0,0 +1,221 @@
+# Copyright Iris contributors
+#
+# This file is part of Iris and is released under the LGPL license.
+# See COPYING and COPYING.LESSER in the root of the repository for full
+# licensing details.
+
+import warnings
+
+import numpy as np
+import shapely
+import shapely.errors
+import shapely.geometry as sgeom
+import shapely.ops
+import shapely.prepared as prepped
+
+from iris.exceptions import IrisUserWarning
+
+# ------------------------------------------------------------------------------
+# GLOBAL VARIABLES
+# ------------------------------------------------------------------------------
+
+
+def create_shapefile_mask(
+    geometry, cube, minimum_weight=0.0, shape_coord_system=None
+):
+    """Get the mask array of the intersection between the given shapely geometry and
+    cube.
+
+    Arguments:
+        geometry        : A :class:`shapely.Geometry` object
+
+        cube            : A :class:`iris.cube.Cube`
+                            with x and y coordinates
+        minimum_weight  : A float between 0 and 1 determining what % of a cell
+                            a shape must cover for the cell to remain unmasked.
+                            eg: 0.1 means that at least 10% of the shape overlaps the cell
+                            to be unmasked
+
+    Returns:
+        A numpy array of the shape of the x & y coordinates of the cube, with points to mask equal to True
+    """
+
+    from iris.cube import Cube, CubeList
+
+    try:
+        if geometry.is_valid is False:
+            raise TypeError("Geometry is not a valid Shapely object")
+    except Exception:
+        raise TypeError("Geometry is not a valid Shapely object")
+    if not isinstance(cube, Cube):
+        if isinstance(cube, CubeList):
+            msg = "Received CubeList object rather than Cube - to mask a CubeList iterate over each Cube"
+            raise TypeError(msg)
+        else:
+            msg = "Received non-Cube object where a Cube is expected"
+            raise TypeError(msg)
+    if not minimum_weight or minimum_weight == 0.0:
+        weights = False
+    elif isinstance(
+        geometry,
+        (
+            sgeom.Point,
+            sgeom.LineString,
+            sgeom.LinearRing,
+            sgeom.MultiPoint,
+            sgeom.MultiLineString,
+        ),
+    ):
+        weights = False
+        warnings.warn(
+            "Shape is of invalid type for minimum weight masking, must use a Polygon rather than Line shape.\n Masking based off intersection instead. ",
+            IrisUserWarning,
+        )
+    else:
+        weights = True
+
+    # prepare shape
+    trans_geo = _transform_coord_system(geometry, cube, shape_coord_system)
+    prepped.prep(trans_geo)
+
+    # prepare 2D cube
+    y_name, x_name = _cube_primary_xy_coord_names(cube)
```

Comment:
just tested this, and assigning the coords to variables and switching coord_dims to cube_dims actually decreases performance by a substantial margin (the integration tests go from 19.6 seconds to 36.7 seconds). Unsure why this is, but since that is the case revert to before?

Reply:
I wasn't precise enough.

If you only change the `y_name` and `x_name` objects to be `Coord`s, instead of `str`s, then the rest of the code will definitely be slower. `cube.coord()` will readily work with both `str` and `Coord`, but `Coord` will be much slower - it takes longer to check for full `Coord` equality versus `str` matching.

What I actually meant was assigning `Coord` objects AND removing all subsequent calls to `cube.coord()`, since you have already derived the `Coord` object you need.

E.g.

```diff
- xmod = cube.coord(x_name).units.modulus
+ xmod = x_coord.units.modulus
```

If that syntax is slower then I definitely want to hear about it, as it's what we consider Iris best practice.

Does this make sense?

Reply:
Yes - that is what I did for all the calls, for the reason you outlined (directly referring to the coordinate rather than calling it by name from the cube feels like it should be faster). Yet it wasn't, for reasons I don't understand, so I've not implemented best practice here. I probably need to investigate this further.